File: app/crud/tenants.py
# -*- coding: utf-8 -*-

"""
Talvo Admin API - Tenants CRUD
"""

from sqlalchemy.orm import Session
from datetime import datetime, timedelta
from typing import List, Optional
from fastapi import HTTPException, status

from app.models import Tenant, User, ActivityLog
from app.schemas import TenantCreate, TenantUpdate
from app.core.license import LicenseManager
from app.core.security import get_password_hash


def get_tenants(db: Session, skip: int = 0, limit: int = 100) -> List[Tenant]:
    return db.query(Tenant).filter(
        Tenant.is_deleted == False
    ).offset(skip).limit(limit).all()


def get_tenant(db: Session, tenant_id: int) -> Optional[Tenant]:
    return db.query(Tenant).filter(
        Tenant.id == tenant_id,
        Tenant.is_deleted == False
    ).first()


def create_tenant(db: Session, tenant_data: TenantCreate, user_id: int) -> Tenant:
    existing = db.query(Tenant).filter(
        Tenant.name == tenant_data.name,
        Tenant.is_deleted == False
    ).first()
    
    if existing:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="اسم المستأجر موجود بالفعل"
        )
    
    license_info = LicenseManager.generate_license_key(
        tenant_data.name,
        tenant_data.subscription_days
    )
    
    db_name = LicenseManager.create_db_name(license_info['license_key'])
    expiry_date = datetime.utcnow() + timedelta(days=tenant_data.subscription_days)
    
    tenant = Tenant(
        name=tenant_data.name,
        arabic_name=tenant_data.arabic_name,
        license_key=license_info['license_key'],
        db_name=db_name,
        phone=tenant_data.phone,
        email=tenant_data.email,
        address=tenant_data.address,
        tax_number=tenant_data.tax_number,
        commercial_register=tenant_data.commercial_register,
        currency=tenant_data.currency,
        timezone=tenant_data.timezone,
        is_active=tenant_data.is_active,
        max_users=tenant_data.max_users,
        max_storage_mb=tenant_data.max_storage_mb,
        subscription_plan=tenant_data.subscription_plan,
        subscription_expiry=expiry_date,
        can_manage_products=tenant_data.can_manage_products,
        can_manage_sales=tenant_data.can_manage_sales,
        can_manage_purchases=tenant_data.can_manage_purchases,
        can_manage_inventory=tenant_data.can_manage_inventory,
        can_manage_customers=tenant_data.can_manage_customers,
        can_manage_suppliers=tenant_data.can_manage_suppliers,
        can_manage_employees=tenant_data.can_manage_employees,
        can_manage_reports=tenant_data.can_manage_reports,
        can_manage_settings=tenant_data.can_manage_settings,
        can_manage_backup=tenant_data.can_manage_backup,
        can_export_data=tenant_data.can_export_data,
        can_import_data=tenant_data.can_import_data,
        can_manage_roles=tenant_data.can_manage_roles,
        can_view_audit_log=tenant_data.can_view_audit_log,
    )
    
    db.add(tenant)
    db.commit()
    db.refresh(tenant)
    
    # ✅ إنشاء مستخدم مدير للمستأجر
    admin_password = "admin123"
    admin_user = User(
        username="admin",
        email=tenant.email or f"admin@{tenant.name}.com",
        full_name="مدير النظام",
        password_hash=get_password_hash(admin_password),
        is_admin=True,
        is_system_admin_flag=False,
        is_active=True,
        tenant_id=tenant.id
    )
    
    db.add(admin_user)
    db.commit()
    
    # ✅ تسجيل النشاط
    ActivityLog.log(
        db=db,
        user_id=user_id,
        action="create_tenant",
        module="admin",
        description=f"تم إنشاء مستأجر جديد: {tenant.name} (المفتاح: {tenant.license_key[:8]}...)"
    )
    
    return tenant


def update_tenant(db: Session, tenant_id: int, tenant_data: TenantUpdate, user_id: int) -> Tenant:
    tenant = get_tenant(db, tenant_id)
    if not tenant:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="المستأجر غير موجود"
        )
    
    # ✅ تحديث الحقول
    if tenant_data.name is not None:
        tenant.name = tenant_data.name
    if tenant_data.arabic_name is not None:
        tenant.arabic_name = tenant_data.arabic_name
    if tenant_data.phone is not None:
        tenant.phone = tenant_data.phone
    if tenant_data.email is not None:
        tenant.email = tenant_data.email
    if tenant_data.address is not None:
        tenant.address = tenant_data.address
    if tenant_data.tax_number is not None:
        tenant.tax_number = tenant_data.tax_number
    if tenant_data.commercial_register is not None:
        tenant.commercial_register = tenant_data.commercial_register
    if tenant_data.currency is not None:
        tenant.currency = tenant_data.currency
    if tenant_data.timezone is not None:
        tenant.timezone = tenant_data.timezone
    if tenant_data.is_active is not None:
        tenant.is_active = tenant_data.is_active
    if tenant_data.max_users is not None:
        tenant.max_users = tenant_data.max_users
    if tenant_data.max_storage_mb is not None:
        tenant.max_storage_mb = tenant_data.max_storage_mb
    if tenant_data.subscription_plan is not None:
        tenant.subscription_plan = tenant_data.subscription_plan
    if tenant_data.subscription_days is not None:
        tenant.subscription_expiry = datetime.utcnow() + timedelta(days=tenant_data.subscription_days)
    
    # ✅ تحديث الصلاحيات
    if tenant_data.can_manage_products is not None:
        tenant.can_manage_products = tenant_data.can_manage_products
    if tenant_data.can_manage_sales is not None:
        tenant.can_manage_sales = tenant_data.can_manage_sales
    if tenant_data.can_manage_purchases is not None:
        tenant.can_manage_purchases = tenant_data.can_manage_purchases
    if tenant_data.can_manage_inventory is not None:
        tenant.can_manage_inventory = tenant_data.can_manage_inventory
    if tenant_data.can_manage_customers is not None:
        tenant.can_manage_customers = tenant_data.can_manage_customers
    if tenant_data.can_manage_suppliers is not None:
        tenant.can_manage_suppliers = tenant_data.can_manage_suppliers
    if tenant_data.can_manage_employees is not None:
        tenant.can_manage_employees = tenant_data.can_manage_employees
    if tenant_data.can_manage_reports is not None:
        tenant.can_manage_reports = tenant_data.can_manage_reports
    if tenant_data.can_manage_settings is not None:
        tenant.can_manage_settings = tenant_data.can_manage_settings
    if tenant_data.can_manage_backup is not None:
        tenant.can_manage_backup = tenant_data.can_manage_backup
    if tenant_data.can_export_data is not None:
        tenant.can_export_data = tenant_data.can_export_data
    if tenant_data.can_import_data is not None:
        tenant.can_import_data = tenant_data.can_import_data
    if tenant_data.can_manage_roles is not None:
        tenant.can_manage_roles = tenant_data.can_manage_roles
    if tenant_data.can_view_audit_log is not None:
        tenant.can_view_audit_log = tenant_data.can_view_audit_log
    
    db.commit()
    db.refresh(tenant)
    
    ActivityLog.log(
        db=db,
        user_id=user_id,
        action="update_tenant",
        module="admin",
        description=f"تم تحديث المستأجر: {tenant.name}"
    )
    
    return tenant


def delete_tenant(db: Session, tenant_id: int, force: bool, user_id: int):
    tenant = get_tenant(db, tenant_id)
    if not tenant:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="المستأجر غير موجود"
        )
    
    if tenant.total_users > 0 and not force:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"لا يمكن حذف المستأجر لأنه يحتوي على {tenant.total_users} مستخدمين نشطين"
        )
    
    tenant.is_deleted = True
    db.commit()
    
    ActivityLog.log(
        db=db,
        user_id=user_id,
        action="delete_tenant",
        module="admin",
        description=f"تم حذف المستأجر: {tenant.name}"
    )
    
    return {"success": True, "message": f"تم حذف المستأجر {tenant.name} بنجاح"}


def toggle_status(db: Session, tenant_id: int, user_id: int):
    tenant = get_tenant(db, tenant_id)
    if not tenant:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="المستأجر غير موجود"
        )
    
    tenant.is_active = not tenant.is_active
    db.commit()
    
    status = "نشط" if tenant.is_active else "غير نشط"
    
    ActivityLog.log(
        db=db,
        user_id=user_id,
        action="toggle_tenant_status",
        module="admin",
        description=f"تم تغيير حالة المستأجر {tenant.name} إلى {status}"
    )
    
    return {"success": True, "message": f"تم تغيير حالة المستأجر إلى {status}"}

File: src/controllers/tenants.ts
import { Request, Response } from 'express';
import { PrismaClient } from '@prisma/client';
import { LicenseManager } from '../utils/license';
import { hashPassword } from '../utils/security';

const prisma = new PrismaClient();

export class TenantsController {
  async getAll(req: Request, res: Response) {
    try {
      const { skip = 0, limit = 100 } = req.query;
      
      const tenants = await prisma.tenant.findMany({
        where: { is_deleted: false },
        skip: Number(skip),
        take: Number(limit),
        orderBy: { created_at: 'desc' }
      });
      
      res.json(tenants);
    } catch (error) {
      res.status(500).json({ error: 'Failed to fetch tenants' });
    }
  }

  async getById(req: Request, res: Response) {
    try {
      const { id } = req.params;
      
      const tenant = await prisma.tenant.findFirst({
        where: { id: Number(id), is_deleted: false }
      });
      
      if (!tenant) {
        return res.status(404).json({ error: 'Tenant not found' });
      }
      
      res.json(tenant);
    } catch (error) {
      res.status(500).json({ error: 'Failed to fetch tenant' });
    }
  }

  async create(req: Request, res: Response) {
    try {
      const data = req.body;
      
      // ✅ Check if tenant name exists
      const existing = await prisma.tenant.findFirst({
        where: { name: data.name, is_deleted: false }
      });
      
      if (existing) {
        return res.status(400).json({ error: 'Tenant name already exists' });
      }
      
      // ✅ Generate license key
      const licenseInfo = LicenseManager.generateLicenseKey(
        data.name,
        data.subscription_days || 365
      );
      
      const dbName = LicenseManager.createDbName(licenseInfo.license_key);
      
      // ✅ Calculate expiry date
      const expiryDate = new Date();
      expiryDate.setDate(expiryDate.getDate() + (data.subscription_days || 365));
      
      // ✅ Create tenant
      const tenant = await prisma.tenant.create({
        data: {
          name: data.name,
          arabic_name: data.arabic_name,
          license_key: licenseInfo.license_key,
          db_name: dbName,
          phone: data.phone,
          email: data.email,
          address: data.address,
          tax_number: data.tax_number,
          commercial_register: data.commercial_register,
          currency: data.currency || 'EGP',
          timezone: data.timezone || 'Africa/Cairo',
          is_active: data.is_active ?? true,
          max_users: data.max_users || 5,
          max_storage_mb: data.max_storage_mb || 100,
          subscription_plan: data.subscription_plan || 'basic',
          subscription_expiry: expiryDate,
          can_manage_products: data.can_manage_products ?? true,
          can_manage_sales: data.can_manage_sales ?? true,
          can_manage_purchases: data.can_manage_purchases ?? true,
          can_manage_inventory: data.can_manage_inventory ?? true,
          can_manage_customers: data.can_manage_customers ?? true,
          can_manage_suppliers: data.can_manage_suppliers ?? true,
          can_manage_employees: data.can_manage_employees ?? true,
          can_manage_reports: data.can_manage_reports ?? true,
          can_manage_settings: data.can_manage_settings ?? false,
          can_manage_backup: data.can_manage_backup ?? false,
          can_export_data: data.can_export_data ?? false,
          can_import_data: data.can_import_data ?? false,
          can_manage_roles: data.can_manage_roles ?? false,
          can_view_audit_log: data.can_view_audit_log ?? false
        }
      });
      
      // ✅ Create admin user
      const adminPassword = 'admin123';
      const adminUser = await prisma.user.create({
        data: {
          username: 'admin',
          email: data.email || `admin@${data.name}.com`,
          full_name: 'مدير النظام',
          password_hash: hashPassword(adminPassword),
          is_admin: true,
          is_system_admin_flag: false,
          is_active: true,
          tenant_id: tenant.id
        }
      });
      
      res.status(201).json({
        success: true,
        tenant,
        license_key: tenant.license_key,
        admin_password: adminPassword,
        message: `تم إنشاء المستأجر ${tenant.name} بنجاح`
      });
    } catch (error) {
      console.error('Create tenant error:', error);
      res.status(500).json({ error: 'Failed to create tenant' });
    }
  }

  async update(req: Request, res: Response) {
    try {
      const { id } = req.params;
      const data = req.body;
      
      // ✅ Check if tenant exists
      const tenant = await prisma.tenant.findFirst({
        where: { id: Number(id), is_deleted: false }
      });
      
      if (!tenant) {
        return res.status(404).json({ error: 'Tenant not found' });
      }
      
      // ✅ Update subscription expiry if days provided
      let subscriptionExpiry = tenant.subscription_expiry;
      if (data.subscription_days) {
        const newExpiry = new Date();
        newExpiry.setDate(newExpiry.getDate() + data.subscription_days);
        subscriptionExpiry = newExpiry;
      }
      
      // ✅ Update tenant
      const updated = await prisma.tenant.update({
        where: { id: Number(id) },
        data: {
          name: data.name ?? tenant.name,
          arabic_name: data.arabic_name ?? tenant.arabic_name,
          phone: data.phone ?? tenant.phone,
          email: data.email ?? tenant.email,
          address: data.address ?? tenant.address,
          tax_number: data.tax_number ?? tenant.tax_number,
          commercial_register: data.commercial_register ?? tenant.commercial_register,
          currency: data.currency ?? tenant.currency,
          timezone: data.timezone ?? tenant.timezone,
          is_active: data.is_active ?? tenant.is_active,
          max_users: data.max_users ?? tenant.max_users,
          max_storage_mb: data.max_storage_mb ?? tenant.max_storage_mb,
          subscription_plan: data.subscription_plan ?? tenant.subscription_plan,
          subscription_expiry: subscriptionExpiry,
          can_manage_products: data.can_manage_products ?? tenant.can_manage_products,
          can_manage_sales: data.can_manage_sales ?? tenant.can_manage_sales,
          can_manage_purchases: data.can_manage_purchases ?? tenant.can_manage_purchases,
          can_manage_inventory: data.can_manage_inventory ?? tenant.can_manage_inventory,
          can_manage_customers: data.can_manage_customers ?? tenant.can_manage_customers,
          can_manage_suppliers: data.can_manage_suppliers ?? tenant.can_manage_suppliers,
          can_manage_employees: data.can_manage_employees ?? tenant.can_manage_employees,
          can_manage_reports: data.can_manage_reports ?? tenant.can_manage_reports,
          can_manage_settings: data.can_manage_settings ?? tenant.can_manage_settings,
          can_manage_backup: data.can_manage_backup ?? tenant.can_manage_backup,
          can_export_data: data.can_export_data ?? tenant.can_export_data,
          can_import_data: data.can_import_data ?? tenant.can_import_data,
          can_manage_roles: data.can_manage_roles ?? tenant.can_manage_roles,
          can_view_audit_log: data.can_view_audit_log ?? tenant.can_view_audit_log
        }
      });
      
      res.json({
        success: true,
        tenant: updated,
        message: `تم تحديث المستأجر ${updated.name} بنجاح`
      });
    } catch (error) {
      console.error('Update tenant error:', error);
      res.status(500).json({ error: 'Failed to update tenant' });
    }
  }

  async delete(req: Request, res: Response) {
    try {
      const { id } = req.params;
      const { force } = req.query;
      
      // ✅ Check if tenant exists
      const tenant = await prisma.tenant.findFirst({
        where: { id: Number(id), is_deleted: false }
      });
      
      if (!tenant) {
        return res.status(404).json({ error: 'Tenant not found' });
      }
      
      // ✅ Check if tenant has users
      if (tenant.total_users > 0 && !force) {
        return res.status(400).json({
          error: `لا يمكن حذف المستأجر لأنه يحتوي على ${tenant.total_users} مستخدمين نشطين`
        });
      }
      
      // ✅ Soft delete
      await prisma.tenant.update({
        where: { id: Number(id) },
        data: { is_deleted: true }
      });
      
      res.json({
        success: true,
        message: `تم حذف المستأجر ${tenant.name} بنجاح`
      });
    } catch (error) {
      console.error('Delete tenant error:', error);
      res.status(500).json({ error: 'Failed to delete tenant' });
    }
  }

  async toggleStatus(req: Request, res: Response) {
    try {
      const { id } = req.params;
      
      // ✅ Check if tenant exists
      const tenant = await prisma.tenant.findFirst({
        where: { id: Number(id), is_deleted: false }
      });
      
      if (!tenant) {
        return res.status(404).json({ error: 'Tenant not found' });
      }
      
      // ✅ Toggle status
      const updated = await prisma.tenant.update({
        where: { id: Number(id) },
        data: { is_active: !tenant.is_active }
      });
      
      const status = updated.is_active ? 'نشط' : 'غير نشط';
      
      res.json({
        success: true,
        message: `تم تغيير حالة المستأجر إلى ${status}`
      });
    } catch (error) {
      console.error('Toggle status error:', error);
      res.status(500).json({ error: 'Failed to toggle status' });
    }
  }
}